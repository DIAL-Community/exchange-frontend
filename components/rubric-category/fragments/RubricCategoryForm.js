import React, { useCallback, useContext, useState } from 'react'
import { useRouter } from 'next/router'
import { Controller, useForm } from 'react-hook-form'
import { FaSpinner } from 'react-icons/fa'
import { useIntl } from 'react-intl'
import { useMutation } from '@apollo/client'
import { ToastContext } from '../../../lib/ToastContext'
import { HtmlEditor } from '../../shared/form/HtmlEditor'
import Input from '../../shared/form/Input'
import ValidationError from '../../shared/form/ValidationError'
import { CREATE_RUBRIC_CATEGORY } from '../../shared/mutation/rubricCategory'

const RubricCategoryForm = React.memo(({ rubricCategory }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const slug = rubricCategory?.slug ?? ''

  const router = useRouter()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const { locale } = router

  const [updateRubricCategory, { reset }] = useMutation(CREATE_RUBRIC_CATEGORY, {
    onCompleted: (data) => {
      const { createRubricCategory: response } = data
      if (response?.rubricCategory && response?.errors?.length === 0) {
        setMutating(false)
        showSuccessMessage(
          format('toast.submit.success', { entity: format('ui.rubricCategory.label') }),
          () => router.push(`/${router.locale}/rubric-categories/${response?.rubricCategory?.slug}`)
        )
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('ui.rubricCategory.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('ui.rubricCategory.label') }))
      setMutating(false)
      reset()
    }
  })

  const { handleSubmit, register, control, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: rubricCategory?.name,
      weight: rubricCategory?.weight,
      description: rubricCategory?.rubricCategoryDescription?.description
    }
  })

  const doUpsert = async (data) => {
    setMutating(true)
    const { name, description } = data
    const weight = parseFloat(data.weight)

    const variables = {
      name,
      slug,
      weight,
      description
    }

    updateRubricCategory({
      variables,
      context: {
        headers: {
          'Accept-Language': locale
        }
      }
    })
  }

  const cancelForm = () => {
    setReverting(true)
    router.push(`/rubric-categories/${rubricCategory?.slug ?? ''}`)
  }

  return (
    <form onSubmit={handleSubmit(doUpsert)}>
      <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-plum'>
        <div className='flex flex-col gap-y-6 text-sm'>
          <div className='text-xl font-semibold'>
            {rubricCategory
              ? format('app.editEntity', { entity: rubricCategory.name })
              : `${format('app.createNew')} ${format('ui.rubricCategory.label')}`
            }
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='text-dial-sapphire required-field' htmlFor='name'>
              {format('app.name')}
            </label>
            <Input
              {...register('name', { required: format('validation.required') })}
              id='name'
              placeholder={format('app.name')}
              isInvalid={errors.name}
            />
            {errors.name && <ValidationError value={errors.name?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='text-dial-sapphire required-field' htmlFor='weight'>
              {format('ui.rubricCategory.weight')}
            </label>
            <Input
              {...register('weight', { required: format('validation.required') })}
              id='weight'
              type='number'
              step='0.1'
              placeholder={format('ui.rubricCategory.weight')}
              isInvalid={errors.weight}
            />
            {errors.weight && <ValidationError value={errors.weight?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label id='description' className='text-dial-sapphire required-field'>
              {format('app.description')}
            </label>
            <Controller
              name='description'
              control={control}
              render={({ field: { value, onChange } }) => (
                <HtmlEditor
                  labelledBy='description'
                  onChange={onChange}
                  initialContent={value}
                  placeholder={format('app.description')}
                />
              )}
              rules={{ required: format('validation.required') }}
            />
            {errors.description && <ValidationError value={errors.description?.message} />}
          </div>
          <div className='flex flex-wrap text-base mt-6 gap-3'>
            <button
              type='submit'
              className='submit-button'
              disabled={mutating || reverting}
            >
              {`${format('app.submit')} ${format('ui.rubricCategory.label')}`}
              {mutating && <FaSpinner className='spinner ml-3' />}
            </button>
            <button
              type='button'
              className='cancel-button'
              disabled={mutating || reverting}
              onClick={cancelForm}
            >
              {format('app.cancel')}
              {reverting && <FaSpinner className='spinner ml-3' />}
            </button>
          </div>
        </div>
      </div>
    </form>
  )
})

RubricCategoryForm.displayName = 'RubricCategoryForm'

export default RubricCategoryForm
