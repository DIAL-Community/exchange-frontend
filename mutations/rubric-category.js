import { gql } from '@apollo/client'

export const CREATE_RUBRIC_CATEGORY = gql`
  mutation CreateRubricCategory (
    $name: String!
    $slug: String!
    $weight: Float!
    $description: String!
  ) {
    createRubricCategory(
      name: $name
      slug: $slug
      weight: $weight
      description: $description
    ) {
      rubricCategory{
        id
        slug
      }
      errors
    }
  }
`

export const DELETE_RUBRIC_CATEGORY = gql`
  mutation DeleteRubricCategory ($id: ID!) {
    deleteRubricCategory(id: $id) {
      errors
    }
  }
`

export const UPDATE_RUBRIC_CATEGORY_INDICATORS = gql`
  mutation UpdateRubricCategoryIndicators (
    $categoryIndicatorSlugs: [String!]!
    $rubricCategorySlug: String!
  ) {
    updateRubricCategoryIndicators (
      categoryIndicatorSlugs: $categoryIndicatorSlugs
      rubricCategorySlug: $rubricCategorySlug
    ) {
      rubricCategory {
        categoryIndicators {
          slug
          name
          rubricCategoryId
        }
      } 
      errors
    }
  }
`
