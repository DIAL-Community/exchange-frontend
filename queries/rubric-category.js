import { gql } from '@apollo/client'

export const RUBRIC_CATEGORIES_LIST_QUERY = gql`
  query RubricCategories($search: String!) { 
    rubricCategories(search: $search) {
      nodes {
        name
        slug
        weight
      }
      totalCount
    }
  }
`

export const RUBRIC_CATEGORY_QUERY = gql`
  query RubricCategory ($slug: String!) {
    rubricCategory (slug: $slug) {
      name
      slug
      weight
      rubricCategoryDescription {
        description
      }
    }
  }
`
