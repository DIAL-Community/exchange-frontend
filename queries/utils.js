const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
