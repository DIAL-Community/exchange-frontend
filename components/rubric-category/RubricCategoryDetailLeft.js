import Bookmark from '../shared/common/Bookmark'
import Comment from '../shared/common/Comment'
import Share from '../shared/common/Share'
import { ObjectType } from '../utils/constants'
import RubricCategoryDetailHeader from './fragments/RubricCategoryDetailHeader'
import RubricCategoryDetailNav from './fragments/RubricCategoryDetailNav'

const RubricCategoryDetailLeft = ({ scrollRef, rubricCategory }) => {
  return (
    <div className='bg-dial-slate-100 lg:h-full'>
      <div className='flex flex-col gap-y-3 px-4 lg:px-6 lg:py-3'>
        <RubricCategoryDetailHeader rubricCategory={rubricCategory} />
        <hr className='border-b border-dial-slate-200' />
        <RubricCategoryDetailNav rubricCategory={rubricCategory} scrollRef={scrollRef} />
        <hr className='border-b border-dial-slate-200' />
        <div className='hidden lg:flex flex-col gap-y-3'>
          <Bookmark object={rubricCategory} objectType={ObjectType.SDG} />
          <hr className='border-b border-dial-slate-200' />
          <Share />
          <hr className='border-b border-dial-slate-200' />
          <Comment entityKey={'ui.rubricCategory.label'} scrollRef={scrollRef} />
          <hr className='border-b border-dial-slate-200' />
        </div>
      </div>
    </div>
  )
}

export default RubricCategoryDetailLeft
