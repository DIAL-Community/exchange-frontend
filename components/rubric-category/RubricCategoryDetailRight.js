import { forwardRef, useCallback, useImperativeHandle, useRef } from 'react'
import { useIntl } from 'react-intl'
import { useQuery } from '@apollo/client'
import { EDITING_POLICY_SLUG, GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import CategoryIndicatorCard from '../category-indicator/CategoryIndicatorCard'
import CommentsSection from '../shared/comment/CommentsSection'
import Bookmark from '../shared/common/Bookmark'
import Share from '../shared/common/Share'
import CreateButton from '../shared/form/CreateButton'
import EditButton from '../shared/form/EditButton'
import { HtmlViewer } from '../shared/form/HtmlViewer'
import { RUBRIC_CATEGORY_QUERY } from '../shared/query/rubricCategory'
import { DisplayType, ObjectType } from '../utils/constants'
import DeleteRubricCategory from './buttons/DeleteRubricCategory'

const RubricCategoryDetailRight = forwardRef(({ rubricCategory }, ref) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const descRef = useRef()
  const categoryIndicatorRef = useRef()
  const commentsSectionRef = useRef()
  useImperativeHandle(
    ref,
    () => [
      { value: 'ui.common.detail.description', ref: descRef },
      { value: 'categoryIndicator.header', ref: categoryIndicatorRef },
      { value: 'ui.comment.label', ref: commentsSectionRef }
    ],
    []
  )

  const editPath = `${rubricCategory.slug}/edit`

  let editingAllowed = true
  const { error } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug: EDITING_POLICY_SLUG },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.EDITING
      }
    }
  })

  if (error) {
    editingAllowed = false
  }

  return (
    <div className='px-4 lg:px-0 py-4 lg:py-6'>
      <div className='flex flex-col gap-y-3'>
        {editingAllowed && (
          <div className='flex gap-x-3 ml-auto'>
            <EditButton type='link' href={editPath} />
            <DeleteRubricCategory rubricCategory={rubricCategory} />
          </div>
        )}
        <div className='text-xl font-semibold text-dial-plum py-3' ref={descRef}>
          {format('ui.common.detail.description')}
        </div>
        <div className='block'>
          <HtmlViewer
            initialContent={rubricCategory?.rubricCategoryDescription?.description}
            editorId='rubric-category-description'
          />
        </div>
        <hr className='border-b border-dial-blue-chalk my-3'/>
        <div className='flex flex-col gap-y-3'>
          <div className='flex flex-row pb-3' ref={categoryIndicatorRef}>
            <div className='text-xl font-semibold text-dial-blueberry '>
              {format('categoryIndicator.header')}
            </div>
            {editingAllowed &&
              <CreateButton
                type='link'
                className='ml-auto'
                label={format('app.create')}
                href={
                  `/rubric-categories/${rubricCategory.slug}` +
                  '/category-indicators/create'
                }
              />
            }
          </div>
          <div className='flex flex-col gap-y-3'>
            {rubricCategory?.categoryIndicators?.map((categoryIndicator, index) =>
              <div key={index}>
                <CategoryIndicatorCard
                  rubricCategory={rubricCategory}
                  categoryIndicator={categoryIndicator}
                  displayType={DisplayType.SMALL_CARD}
                />
              </div>
            )}
          </div>
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='lg:hidden flex flex-col gap-y-3'>
          <Bookmark object={rubricCategory} objectType={ObjectType.RUBRIC_CATEGORY} />
          <hr className='border-b border-dial-slate-200'/>
          <Share />
          <hr className='border-b border-dial-slate-200'/>
        </div>
        <CommentsSection
          commentsSectionRef={commentsSectionRef}
          objectId={rubricCategory.id}
          objectType={ObjectType.RUBRIC_CATEGORY}
        />
      </div>
    </div>
  )
})

RubricCategoryDetailRight.displayName = 'RubricCategoryDetailRight'

export default RubricCategoryDetailRight
