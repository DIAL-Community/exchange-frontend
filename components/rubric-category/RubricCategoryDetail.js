import { useEffect, useRef, useState } from 'react'
import { useApolloClient, useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import Breadcrumb from '../shared/Breadcrumb'
import { handleLoadingQuery, handleMissingData, handleQueryError } from '../shared/GraphQueryHandler'
import { RUBRIC_CATEGORY_POLICY_QUERY, RUBRIC_CATEGORY_QUERY } from '../shared/query/rubricCategory'
import { fetchOperationPolicies } from '../utils/policy'
import RubricCategoryDetailLeft from './RubricCategoryDetailLeft'
import RubricCategoryDetailRight from './RubricCategoryDetailRight'

const RubricCategoryDetail = ({ categorySlug }) => {
  const scrollRef = useRef(null)
  const client = useApolloClient()

  const [editingAllowed, setEditingAllowed] = useState(false)
  const [deletingAllowed, setDeletingAllowed] = useState(false)

  const { loading, error, data } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug: categorySlug },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  useEffect(() => {
    fetchOperationPolicies(
      client,
      RUBRIC_CATEGORY_POLICY_QUERY,
      ['editing', 'deleting']
    ).then(policies => {
      setEditingAllowed(policies['editing'])
      setDeletingAllowed(policies['deleting'])
    })
  }, [client])

  if (loading) {
    return handleLoadingQuery()
  } else if (error) {
    return handleQueryError(error)
  } else if (!data?.rubricCategory) {
    return handleMissingData()
  }

  const { rubricCategory } = data

  const slugNameMapping = (() => {
    const map = {}
    map[rubricCategory.slug] = rubricCategory.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping}/>
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3 shrink-0'>
          <RubricCategoryDetailLeft
            scrollRef={scrollRef}
            rubricCategory={rubricCategory}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
        <div className='lg:basis-2/3 shrink-0'>
          <RubricCategoryDetailRight
            ref={scrollRef}
            rubricCategory={rubricCategory}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
      </div>
    </div>
  )
}

export default RubricCategoryDetail
