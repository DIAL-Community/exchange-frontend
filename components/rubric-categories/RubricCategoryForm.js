import React, { useState, useCallback, useMemo, useContext } from 'react'
import { useRouter } from 'next/router'
import { useMutation } from '@apollo/client'
import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa'
import { Controller, useForm } from 'react-hook-form'
import Breadcrumb from '../shared/breadcrumb'
import { HtmlEditor } from '../shared/HtmlEditor'
import Input from '../shared/Input'
import { ToastContext } from '../../lib/ToastContext'
import ValidationError from '../shared/ValidationError'
import { useUser } from '../../lib/hooks'
import { CREATE_RUBRIC_CATEGORY } from '../../mutations/rubric-category'

const RubricCategoryForm = React.memo(({ rubricCategory }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const slug = rubricCategory?.slug ?? ''

  const router = useRouter()
  const { user } = useUser()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const { showToast } = useContext(ToastContext)

  const { locale } = router

  const [updateRubricCategory, { reset }] = useMutation(CREATE_RUBRIC_CATEGORY, {
    onCompleted: (data) => {
      const { createRubricCategory: response } = data
      if (response?.rubricCategory && response?.errors?.length === 0) {
        setMutating(false)
        showToast(
          format('toast.rubric-category.submit.success'),
          'success',
          'top-center',
          1000,
          null,
          () => router.push(`/${router.locale}/rubric_categories/${response?.rubricCategory?.slug}`)
        )
      } else {
        setMutating(false)
        showToast(format('toast.rubric-category.submit.failure'), 'error', 'top-center')
        reset()
      }
    },
    onError: (error) => {
      setMutating(false)
      showToast(
        <div className='flex flex-col'>
          <span>{format('toast.rubric-category.submit.failure')}</span>
          <span>{error?.message}</span>
        </div>,
        'error',
        'top-center'
      )
      reset()
    }
  })

  const { handleSubmit, register, control, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: rubricCategory?.name,
      weight: rubricCategory?.weight,
      description: rubricCategory?.rubricCategoryDescription?.description
    }
  })

  const slugNameMapping = useMemo(() => {
    const map = {
      create: format('app.create'),
      edit: format('app.edit')
    }

    if (rubricCategory) {
      map[rubricCategory.slug] = rubricCategory.name
    }

    return map
  }, [rubricCategory, format])

  const doUpsert = async (data) => {
    if (user) {
      setMutating(true)
      const { userEmail, userToken } = user
      const { name, description } = data
      const weight = parseFloat(data.weight)

      const variables = {
        name,
        slug,
        weight,
        description
      }

      updateRubricCategory({
        variables,
        context: {
          headers: {
            'Accept-Language': locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)
    router.push(`/rubric_categories/${rubricCategory?.slug ?? ''}`)
  }

  return (
    <div className='flex flex-col'>
      <div className='hidden lg:block px-8'>
        <Breadcrumb slugNameMapping={slugNameMapping} />
      </div>
      <div className='pb-8 px-8'>
        <div id='content' className='sm:px-0 max-w-full mx-auto'>
          <form onSubmit={handleSubmit(doUpsert)}>
            <div className='bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3'>
              <div className='text-2xl font-semibold text-dial-sapphire pb-4'>
                {rubricCategory
                  ? format('app.editEntity', { entity: rubricCategory.name })
                  : `${format('app.createNew')} ${format('rubric-category.label')}`
                }
              </div>
              <div className='flex flex-col lg:flex-row gap-4'>
                <div className='w-full lg:w-1/2 flex flex-col gap-y-3'>
                  <div className='form-field-wrapper' data-testid='rubric-category-name'>
                    <label className='form-field-label required-field' htmlFor='name'>
                      {format('app.name')}
                    </label>
                    <Input
                      {...register('name', { required: format('validation.required') })}
                      id='name'
                      placeholder={format('app.name')}
                      isInvalid={errors.name}
                    />
                    {errors.name && <ValidationError value={errors.name?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='rubric-category-weight'>
                    <label className='form-field-label required-field' htmlFor='weight'>
                      {format('rubric-category.weight')}
                    </label>
                    <Input
                      {...register('weight', { required: format('validation.required') })}
                      id='weight'
                      type='number'
                      step='0.1'
                      placeholder={format('rubric-category.weight')}
                      isInvalid={errors.weight}
                    />
                    {errors.weight && <ValidationError value={errors.weight?.message} />}
                  </div>
                </div>
                <div className='w-full lg:w-1/2'>
                  <div className='block flex flex-col gap-y-2' data-testid='rubric-category-description'>
                    <label className='form-field-label required-field'>
                      {format('app.description')}
                    </label>
                    <Controller
                      name='description'
                      control={control}
                      render={({ field: { value, onChange } }) => (
                        <HtmlEditor
                          editorId='description-editor'
                          onChange={onChange}
                          initialContent={value}
                          placeholder={format('app.description')}
                          isInvalid={errors.description}
                        />
                      )}
                      rules={{ required: format('validation.required') }}
                    />
                    {errors.description && <ValidationError value={errors.description?.message} />}
                  </div>
                </div>
              </div>
              <div className='flex flex-wrap text-xl mt-8 gap-3'>
                <button
                  type='submit'
                  className='submit-button'
                  disabled={mutating || reverting}
                  data-testid='submit-button'
                >
                  {`${format('app.submit')} ${format('rubric-category.label')}`}
                  {mutating && <FaSpinner className='spinner ml-3' />}
                </button>
                <button
                  type='button'
                  className='cancel-button'
                  disabled={mutating || reverting}
                  onClick={cancelForm}
                >
                  {format('app.cancel')}
                  {reverting && <FaSpinner className='spinner ml-3' />}
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  )
})

RubricCategoryForm.displayName = 'RubricCategoryForm'

export default RubricCategoryForm
