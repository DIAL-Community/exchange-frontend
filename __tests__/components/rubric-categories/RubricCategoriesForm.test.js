import { fireEvent, screen, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import userEvent from '@testing-library/user-event'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import { render, waitForAllEffects } from '../../test-utils'
import { CREATE_RUBRIC_CATEGORY } from '../../../mutations/rubric-category'
import RubricCategoryForm from '../../../components/rubric-categories/RubricCategoryForm'
import { mockNextAuthUseSession, mockNextUseRouter, statuses } from '../../utils/nextMockImplementation'
import { rubricCategory, createRubricCategorySuccess, createRubricCategoryFailure } from './data/RubricCategoryForm'

mockNextUseRouter()
describe('Unit tests for RubricCategoryForm component.', () => {
  const RUBRIC_CATEGORY_NAME_TEST_ID = 'rubric-category-name'
  const RUBRIC_CATEGORY_WEIGHT_TEST_ID = 'rubric-category-weight'
  const RUBRIC_CATEGORY_DESCRIPTION_TEST_ID = 'rubric-category-description'
  const SUBMIT_BUTTON_TEST_ID = 'submit-button'
  const REQUIRED_FIELD_MESSAGE = 'This field is required'

  test('Should render RubricCategoryForm component for admin user.', async () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
    const { container } = render(
      <CustomMockedProvider>
        <RubricCategoryForm />
      </CustomMockedProvider>
    )
    await waitForAllEffects()
    expect(container).toMatchSnapshot()
  })

  test('Should show validation errors for mandatory fields and hide them on input value change.', async () => {
    const user = userEvent.setup()
    mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
    const { getByTestId } = render(
      <CustomMockedProvider>
        <RubricCategoryForm />
      </CustomMockedProvider>
    )
    await waitForAllEffects()
    await act(async () => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(RUBRIC_CATEGORY_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_WEIGHT_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test rubric category name')
    expect(getByTestId(RUBRIC_CATEGORY_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    await act(async () => waitFor(() => {
      user.clear(screen.getByLabelText(/Name/))
    }))
    expect(getByTestId(RUBRIC_CATEGORY_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test rubric category name 2')
    expect(getByTestId(RUBRIC_CATEGORY_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_WEIGHT_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Weight/), '1')
    expect(getByTestId(RUBRIC_CATEGORY_WEIGHT_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await act(async () => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(RUBRIC_CATEGORY_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(RUBRIC_CATEGORY_WEIGHT_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
  })

  describe('Should display toast on submit -', () => {
    test('Success.', async () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
      const mockCreateRubricCategory = generateMockApolloData(
        CREATE_RUBRIC_CATEGORY,
        {
          name: 'Test rubric category',
          slug: 'test_rubric_category',
          weight: 1,
          description: 'RC description'
        },
        null,
        createRubricCategorySuccess
      )
      const { container, getByTestId } = render(
        <CustomMockedProvider mocks={[mockCreateRubricCategory]} addTypename={false}>
          <RubricCategoryForm rubricCategory={rubricCategory} />
        </CustomMockedProvider>
      )

      await waitForAllEffects()
      await act(async () => {
        fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID))
      })
      await screen.findByText('Rubric Category submitted successfully')
      expect(container).toMatchSnapshot()
    })

    test('Failure.', async () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
      const mockCreateRubricCategory = generateMockApolloData(
        CREATE_RUBRIC_CATEGORY,
        {
          name: 'Test rubric category',
          slug: 'test_rubric_category',
          weight: 1,
          description: 'RC description'
        },
        createRubricCategoryFailure
      )
      const { container, getByTestId } = render(
        <CustomMockedProvider mocks={[mockCreateRubricCategory]}>
          <RubricCategoryForm rubricCategory={rubricCategory} />
        </CustomMockedProvider>
      )

      await waitForAllEffects()
      await act(async () => {
        fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID))
      })
      await screen.findByText('Rubric Category submission failed')
      expect(container).toMatchSnapshot()
    })
  })
})
