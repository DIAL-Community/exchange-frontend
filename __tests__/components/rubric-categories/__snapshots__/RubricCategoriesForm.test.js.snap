// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for RubricCategoryForm component. Should render RubricCategoryForm component for admin user. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col"
  >
    <div
      class="hidden lg:block px-8"
    >
      <div
        class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
      >
        <a
          class="inline text-dial-blue h5"
          href="/"
        >
          Home
        </a>
        <div
          class="inline h5"
        >
           &gt; 
          <a
            class="text-dial-gray-dark"
            href="/"
          />
        </div>
      </div>
    </div>
    <div
      class="pb-8 px-8"
    >
      <div
        class="sm:px-0 max-w-full mx-auto"
        id="content"
      >
        <form>
          <div
            class="bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3"
          >
            <div
              class="text-2xl font-bold text-dial-blue pb-4"
            >
              Create New Rubric Category
            </div>
            <div
              class="flex flex-col lg:flex-row gap-4"
            >
              <div
                class="w-full lg:w-1/2 flex flex-col gap-y-3"
              >
                <div
                  class="form-field-wrapper"
                  data-testid="rubric-category-name"
                >
                  <label
                    class="form-field-label required-field"
                    for="name"
                  >
                    Name
                  </label>
                  <input
                    class="w-full"
                    id="name"
                    name="name"
                    placeholder="Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                  data-testid="rubric-category-weight"
                >
                  <label
                    class="form-field-label required-field"
                    for="weight"
                  >
                    Weight
                  </label>
                  <input
                    class="w-full"
                    id="weight"
                    name="weight"
                    placeholder="Weight"
                    step="0.1"
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                class="w-full lg:w-1/2"
              >
                <div
                  class="block flex flex-col gap-y-2"
                  data-testid="rubric-category-description"
                >
                  <label
                    class="form-field-label required-field"
                  >
                    Description
                  </label>
                  <div
                    class="relative"
                  >
                    <svg
                      class="absolute text-lg inset-x-1/2 top-10 spinner"
                      fill="currentColor"
                      height="2em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="2em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                      />
                    </svg>
                    <div
                      class="htmlEditor"
                    >
                      <textarea
                        id="description-editor"
                        style="visibility: hidden;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-wrap text-xl mt-8 gap-3"
            >
              <button
                class="submit-button"
                data-testid="submit-button"
                type="submit"
              >
                Submit Rubric Category
              </button>
              <button
                class="cancel-button"
                type="button"
              >
                Cancel
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
